File: src/infra/crypto/src/bnet_srp3.cpp
// SPDX-License-Identifier: GPL-2.0-or-later

#include "infra/crypto/bnet_srp3.hpp"

#include <algorithm>
#include <array>
#include <cctype>
#include <cstring>
#include <random>
#include <stdexcept>
#include <vector>

#include "infra/crypto/bnet_hash.hpp"

namespace pvpgn::v3::infra::crypto {

namespace {

// Wire bytes of the SRP-3 modulus (Warcraft III); same as legacy
// `bnetsrp3_N`.
constexpr std::array<std::uint8_t, 32> kModulusBytes{
    0xF8, 0xFF, 0x1A, 0x8B, 0x61, 0x99, 0x18, 0x03,
    0x21, 0x86, 0xB6, 0x8C, 0xA0, 0x92, 0xB5, 0x55,
    0x7E, 0x97, 0x6C, 0x78, 0xC7, 0x32, 0x12, 0xD9,
    0x12, 0x16, 0xF6, 0x65, 0x85, 0x23, 0xC7, 0x87};

constexpr std::array<std::uint8_t, 20> kIBytes{
    0xF8, 0x01, 0x8C, 0xF0, 0xA4, 0x25, 0xBA, 0x8B,
    0xEB, 0x89, 0x58, 0xB1, 0xAB, 0x6B, 0xF9, 0x0A,
    0xED, 0x97, 0x0E, 0x6C};

constexpr std::uint8_t kGenerator = 0x2F;

const BigUInt& modulus_singleton()
{
    // Legacy: `BigInt(bnetsrp3_N, 32)` uses the default ctor args
    // `(blockSize=1, bigEndian=true)` -- i.e. pure big-endian.
    static const BigUInt v =
        BigUInt::from_bytes_legacy(kModulusBytes, 1, true);
    return v;
}

const BigUInt& generator_singleton()
{
    static const BigUInt v{static_cast<std::uint32_t>(kGenerator)};
    return v;
}

const BigUInt& i_singleton()
{
    // Legacy: `BigInt(bnetsrp3_I, 32)` uses default ctor args
    // `(blockSize=1, bigEndian=true)` -- pure big-endian.
    static const BigUInt v =
        BigUInt::from_bytes_legacy(kIBytes, 1, true);
    return v;
}

// Toupper that matches legacy `safe_toupper` for ASCII inputs. For
// >0x7f bytes the legacy implementation returns the byte unchanged.
std::string ascii_upper(std::string_view s)
{
    std::string out;
    out.reserve(s.size());
    for (char ch : s) {
        const unsigned char c = static_cast<unsigned char>(ch);
        if (c < 0x80u)
            out.push_back(static_cast<char>(
                std::toupper(static_cast<int>(c))));
        else
            out.push_back(static_cast<char>(c));
    }
    return out;
}

// Generate a 32-byte uniformly random BigUInt. Random bytes are
// drawn from `std::random_device`; the bytes are interpreted as a
// little-endian-block (SRP-3 wire) BigUInt so the value is < 2^256.
BigUInt random_32_byte_biguint()
{
    std::random_device rd;
    std::array<std::uint8_t, 32> bytes{};
    for (auto& b : bytes) {
        b = static_cast<std::uint8_t>(rd() & 0xffu);
    }
    return BigUInt::from_bytes_legacy(bytes, 4, false);
}

// SHA-1 over a contiguous byte buffer (as `std::byte` is what the
// hash API uses, this small helper hides the cast).
BnetDigest sha1_bytes(std::span<const std::uint8_t> data)
{
    auto bytes = std::span<const std::byte>{
        reinterpret_cast<const std::byte*>(data.data()), data.size()};
    return sha1(bytes);
}

BnetDigest sha1_le_bytes(std::span<const std::uint8_t> data)
{
    auto bytes = std::span<const std::byte>{
        reinterpret_cast<const std::byte*>(data.data()), data.size()};
    return sha1_le(bytes);
}

// Pack a BnetDigest (5 host-order uint32) into 20 wire bytes the
// same way as the legacy `t_hash` byte layout: each uint32 is
// stored in native byte order in memory. Match by reinterpreting.
std::array<std::uint8_t, 20> digest_to_bytes(const BnetDigest& d)
{
    std::array<std::uint8_t, 20> out{};
    std::memcpy(out.data(), d.data(), 20);
    return out;
}

}  // namespace

// ----- static accessors -------------------------------------------------

const BigUInt& BnetSrp3::N() { return modulus_singleton(); }
const BigUInt& BnetSrp3::g() { return generator_singleton(); }
const BigUInt& BnetSrp3::I() { return i_singleton(); }

std::array<std::uint8_t, 32> BnetSrp3::n_wire()
{
    std::array<std::uint8_t, 32> out{};
    // Pure big-endian, matching how N was constructed.
    modulus_singleton().to_bytes_legacy(out, 1, true);
    return out;
}

// ----- construction -----------------------------------------------------

BnetSrp3::BnetSrp3(std::string_view username, const BigUInt& salt)
    : username_{ascii_upper(username)},
      password_{},
      a_{std::uint32_t{0u}},
      b_{random_32_byte_biguint() % modulus_singleton()},
      s_{salt}
{
    refresh_raw_salt();
}

BnetSrp3::BnetSrp3(std::string_view username, std::string_view password)
    : username_{ascii_upper(username)},
      password_{ascii_upper(password)},
      a_{random_32_byte_biguint() % modulus_singleton()},
      b_{std::uint32_t{0u}},
      s_{random_32_byte_biguint()}
{
    if (password_.empty()) {
        throw std::invalid_argument{"BnetSrp3: empty password"};
    }
    refresh_raw_salt();
}

void BnetSrp3::set_salt(const BigUInt& s)
{
    s_ = s;
    B_cache_.reset();
    refresh_raw_salt();
}

void BnetSrp3::set_client_private_key(const BigUInt& a)
{
    a_ = a;
    B_cache_.reset();
}

void BnetSrp3::set_server_private_key(const BigUInt& b)
{
    b_ = b;
    B_cache_.reset();
}

void BnetSrp3::refresh_raw_salt()
{
    // legacy: `s.getData(raw_salt, 32)` -- default blockSize=1,
    // bigEndian=true (pure BE).
    s_.to_bytes_legacy(raw_salt_, 1, true);
}

// ----- private helpers --------------------------------------------------

BigUInt BnetSrp3::client_private_key() const
{
    // x = H( s || H(username:password) )
    // First inner hash: little-endian SHA-1 of "USERNAME:PASSWORD"
    std::string userpass;
    userpass.reserve(username_.size() + 1 + password_.size());
    userpass.append(username_);
    userpass.push_back(':');
    userpass.append(password_);
    const auto userpass_hash = sha1_le_bytes(std::span<const std::uint8_t>{
        reinterpret_cast<const std::uint8_t*>(userpass.data()),
        userpass.size()});
    const auto userpass_bytes = digest_to_bytes(userpass_hash);

    std::array<std::uint8_t, 32 + 20> private_value{};
    std::memcpy(private_value.data(), raw_salt_.data(), 32);
    std::memcpy(private_value.data() + 32, userpass_bytes.data(), 20);

    const auto private_hash = sha1_le_bytes(private_value);
    const auto private_bytes = digest_to_bytes(private_hash);

    // Legacy returns BigInt(private_value_hash, 20, 1, false)
    return BigUInt::from_bytes_legacy(private_bytes, 1, false);
}

BigUInt BnetSrp3::scrambler(const BigUInt& B) const
{
    // scrambler = first 32 bits of SHA1(raw_B), interpreted as a
    // host-byte-order uint32 (legacy `*(uint32_t*)hash`).
    std::array<std::uint8_t, 32> raw_B{};
    B.to_bytes_legacy(raw_B, 4, false);
    const auto h = sha1_bytes(raw_B);
    // Legacy reads the first 4 bytes of `t_hash` as uint32. The
    // `t_hash` is `uint32[5]`, so the first uint32 in memory IS
    // h[0] in native byte order. BigInt(uint32_t) just stores the
    // value. We replicate that exactly:
    return BigUInt{static_cast<std::uint32_t>(h[0])};
}

BigUInt BnetSrp3::client_secret(const BigUInt& B) const
{
    const BigUInt x = client_private_key();
    const BigUInt u = scrambler(B);
    const BigUInt& N = modulus_singleton();
    const BigUInt& g = generator_singleton();
    // ((N + B - g^x) ^ (x*u + a)) mod N
    const BigUInt gx   = g.pow_mod(x, N);
    const BigUInt base = (N + B - gx) % N;
    const BigUInt exp  = (x * u) + a_;
    return base.pow_mod(exp, N);
}

BigUInt BnetSrp3::server_secret(const BigUInt& A, const BigUInt& v)
{
    const BigUInt B = server_session_public_key(v);
    const BigUInt u = scrambler(B);
    const BigUInt& N = modulus_singleton();
    const BigUInt base = (A * v.pow_mod(u, N)) % N;
    return base.pow_mod(b_, N);
}

BigUInt BnetSrp3::hash_secret(const BigUInt& secret) const
{
    std::array<std::uint8_t, 32> raw_secret{};
    secret.to_bytes_legacy(raw_secret, 4, false);

    std::array<std::uint8_t, 16> odd{};
    std::array<std::uint8_t, 16> even{};
    for (std::size_t i = 0; i < 16; ++i) {
        odd[i]  = raw_secret[i * 2];
        even[i] = raw_secret[i * 2 + 1];
    }
    const auto odd_h_bytes  = digest_to_bytes(sha1_le_bytes(odd));
    const auto even_h_bytes = digest_to_bytes(sha1_le_bytes(even));

    std::array<std::uint8_t, 40> hashed{};
    for (std::size_t i = 0; i < 20; ++i) {
        hashed[i * 2]     = odd_h_bytes[i];
        hashed[i * 2 + 1] = even_h_bytes[i];
    }
    return BigUInt::from_bytes_legacy(hashed, 1, false);
}

// ----- public API -------------------------------------------------------

BigUInt BnetSrp3::verifier() const
{
    return generator_singleton().pow_mod(client_private_key(),
                                         modulus_singleton());
}

BigUInt BnetSrp3::salt() const { return s_; }

BigUInt BnetSrp3::client_session_public_key() const
{
    return generator_singleton().pow_mod(a_, modulus_singleton());
}

BigUInt BnetSrp3::server_session_public_key(const BigUInt& v)
{
    if (!B_cache_.has_value()) {
        const BigUInt& N = modulus_singleton();
        const BigUInt& g = generator_singleton();
        B_cache_ = (v + g.pow_mod(b_, N)) % N;
    }
    return *B_cache_;
}

BigUInt BnetSrp3::hashed_client_secret(const BigUInt& B) const
{
    return hash_secret(client_secret(B));
}

BigUInt BnetSrp3::hashed_server_secret(const BigUInt& A, const BigUInt& v)
{
    return hash_secret(server_secret(A, v));
}

BigUInt BnetSrp3::client_password_proof(const BigUInt& A, const BigUInt& B,
                                       const BigUInt& K) const
{
    std::array<std::uint8_t, 176> proof_data{};
    const auto username_h = digest_to_bytes(sha1_le_bytes(
        std::span<const std::uint8_t>{
            reinterpret_cast<const std::uint8_t*>(username_.data()),
            username_.size()}));

    // Layout matches legacy:
    //   [0..20)   = I              (20 bytes, blockSize=4)
    //   [20..40)  = H(username)    (20 bytes raw little-endian SHA-1)
    //   [40..72)  = s              (32 bytes, default blockSize=1)
    //   [72..104) = A              (32 bytes, blockSize=4)
    //   [104..136)= B              (32 bytes, blockSize=4)
    //   [136..176)= K              (40 bytes, blockSize=4)
    i_singleton().to_bytes_legacy(
        std::span<std::uint8_t>{proof_data.data() + 0, 20}, 4, false);
    std::memcpy(proof_data.data() + 20, username_h.data(), 20);
    // Legacy: `s.getData(&proofData[40], 32)` -- defaults
    // (blockSize=1, bigEndian=true).
    s_.to_bytes_legacy(
        std::span<std::uint8_t>{proof_data.data() + 40, 32}, 1, true);
    A.to_bytes_legacy(
        std::span<std::uint8_t>{proof_data.data() + 72, 32}, 4, false);
    B.to_bytes_legacy(
        std::span<std::uint8_t>{proof_data.data() + 104, 32}, 4, false);
    K.to_bytes_legacy(
        std::span<std::uint8_t>{proof_data.data() + 136, 40}, 4, false);

    const auto proof_h_bytes =
        digest_to_bytes(sha1_le_bytes(proof_data));
    return BigUInt::from_bytes_legacy(proof_h_bytes, 1, false);
}

BigUInt BnetSrp3::server_password_proof(const BigUInt& A, const BigUInt& M,
                                       const BigUInt& K) const
{
    std::array<std::uint8_t, 92> proof_data{};
    A.to_bytes_legacy(
        std::span<std::uint8_t>{proof_data.data() + 0, 32}, 4, false);
    M.to_bytes_legacy(
        std::span<std::uint8_t>{proof_data.data() + 32, 20}, 4, false);
    K.to_bytes_legacy(
        std::span<std::uint8_t>{proof_data.data() + 52, 40}, 4, false);

    const auto proof_h_bytes =
        digest_to_bytes(sha1_le_bytes(proof_data));
    return BigUInt::from_bytes_legacy(proof_h_bytes, 1, false);
}

}  // namespace pvpgn::v3::infra::crypto

File: tests/diff/bnet_srp3.py
# SPDX-License-Identifier: GPL-2.0-or-later
"""Faithful Python port of Battle.net SRP-3 (WarCraft III / W3XP).

A byte-for-byte re-implementation of the parity-verified C++
`pvpgn::v3::infra::crypto::BnetSrp3` (which is itself bit-compatible with the
original server's `BnetSRP3`). Used by the differential harness to drive the
WAR3/W3XP NLS login (SID_AUTH_ACCOUNTCREATE 0x52 / ACCOUNTLOGON 0x53 / PROOF
0x54) against BOTH the original pvpgn-server and v3.

Correctness is pinned by golden vectors captured from the C++ golden test
(tests/unit/infra/crypto/bnet_srp3_golden_test.cpp) — run this module directly
to self-check:  python3 tests/diff/bnet_srp3.py
"""
import hashlib

# --- SRP-3 constants (Warcraft III) -----------------------------------------
_MODULUS_BYTES = bytes([
    0xF8, 0xFF, 0x1A, 0x8B, 0x61, 0x99, 0x18, 0x03,
    0x21, 0x86, 0xB6, 0x8C, 0xA0, 0x92, 0xB5, 0x55,
    0x7E, 0x97, 0x6C, 0x78, 0xC7, 0x32, 0x12, 0xD9,
    0x12, 0x16, 0xF6, 0x65, 0x85, 0x23, 0xC7, 0x87])
_I_BYTES = bytes([
    0xF8, 0x01, 0x8C, 0xF0, 0xA4, 0x25, 0xBA, 0x8B,
    0xEB, 0x89, 0x58, 0xB1, 0xAB, 0x6B, 0xF9, 0x0A,
    0xED, 0x97, 0x0E, 0x6C])

N = int.from_bytes(_MODULUS_BYTES, "big")   # from_bytes_legacy(.,1,true)
G = 0x2F                                     # generator
I = int.from_bytes(_I_BYTES, "big")          # from_bytes_legacy(.,1,true)


# --- BigUInt legacy block conversions (ported from big_uint.cpp) -------------
def from_bytes_legacy(b: bytes, block: int, big_endian: bool) -> int:
    """Mirror BigUInt::from_bytes_legacy."""
    work = bytearray(b)
    if not big_endian and block > 1:
        for i in range(0, len(work) - block + 1, block):
            work[i:i + block] = work[i:i + block][::-1]
    if big_endian:
        return int.from_bytes(work, "big")
    # !big_endian path packs into 4-byte segments LSB-first, segments
    # little-endian-ordered -> the whole buffer read little-endian.
    return int.from_bytes(work, "little")


def to_bytes_legacy(v: int, n: int, block: int, big_endian: bool) -> bytes:
    """Mirror BigUInt::to_bytes_legacy."""
    out = bytearray(n)
    val = v
    i = 0
    while i + 4 <= n:
        seg = val & 0xFFFFFFFF
        val >>= 32
        out[i] = (seg >> 24) & 0xFF
        out[i + 1] = (seg >> 16) & 0xFF
        out[i + 2] = (seg >> 8) & 0xFF
        out[i + 3] = seg & 0xFF
        i += 4
    if i < n:  # tail: remaining low bytes at the end, big-endian
        tail = n - i
        for bb in range(tail):
            out[n - 1 - bb] = val & 0xFF
            val >>= 8
    if not big_endian and block > 1:
        for k in range(0, n - block + 1, block):
            out[k:k + block] = out[k:k + block][::-1]
    return bytes(out)


# --- SHA-1 helpers -----------------------------------------------------------
# digest_to_bytes(sha1_le(x)) == standard SHA-1 digest (20 bytes, big-endian
# words). digest_to_bytes(sha1(x)) would be the byte-swapped words, but every
# place we need raw bytes uses sha1_le, so H() == hashlib.sha1.
def _H(data: bytes) -> bytes:
    return hashlib.sha1(data).digest()


def _ascii_upper(s: str) -> str:
    return "".join(c.upper() if ord(c) < 0x80 else c for c in s)


class BnetSrp3:
    """Client-side SRP-3 (knows the password). Salt may be pinned for
    determinism; the client private key `a` likewise."""

    def __init__(self, username: str, password: str):
        self.username = _ascii_upper(username)
        self.password = _ascii_upper(password)
        self._salt = 0
        self._a = 0

    # ----- determinism hooks -----
    def set_salt(self, s: int):
        self._salt = s

    def set_client_private_key(self, a: int):
        self._a = a

    # ----- raw salt (32 bytes, blockSize=1 big-endian) -----
    def _raw_salt(self) -> bytes:
        return to_bytes_legacy(self._salt, 32, 1, True)

    # ----- x = H(s || H(USER:PASS)) -----
    def _client_private_key(self) -> int:
        userpass = (self.username + ":" + self.password).encode("latin-1")
        userpass_hash = _H(userpass)                 # 20 bytes
        private_value = self._raw_salt() + userpass_hash  # 52 bytes
        private_hash = _H(private_value)             # 20 bytes
        return from_bytes_legacy(private_hash, 1, False)  # little-endian

    def verifier(self) -> int:
        return pow(G, self._client_private_key(), N)

    def client_session_public_key(self) -> int:
        return pow(G, self._a, N)

    # ----- u = first 32-bit word of standard SHA-1(raw_B) -----
    @staticmethod
    def _scrambler(B: int) -> int:
        raw_B = to_bytes_legacy(B, 32, 4, False)
        return int.from_bytes(_H(raw_B)[0:4], "big")

    def _client_secret(self, B: int) -> int:
        x = self._client_private_key()
        u = self._scrambler(B)
        gx = pow(G, x, N)
        base = (N + B - gx) % N
        exp = x * u + self._a
        return pow(base, exp, N)

    @staticmethod
    def _hash_secret(secret: int) -> int:
        raw = to_bytes_legacy(secret, 32, 4, False)
        odd = bytes(raw[i * 2] for i in range(16))
        even = bytes(raw[i * 2 + 1] for i in range(16))
        odd_h = _H(odd)
        even_h = _H(even)
        hashed = bytearray(40)
        for i in range(20):
            hashed[i * 2] = odd_h[i]
            hashed[i * 2 + 1] = even_h[i]
        return from_bytes_legacy(bytes(hashed), 1, False)

    def hashed_client_secret(self, B: int) -> int:
        return self._hash_secret(self._client_secret(B))

    def client_password_proof(self, A: int, B: int, K: int) -> int:
        pd = bytearray(176)
        pd[0:20] = to_bytes_legacy(I, 20, 4, False)
        pd[20:40] = _H(self.username.encode("latin-1"))
        pd[40:72] = to_bytes_legacy(self._salt, 32, 1, True)
        pd[72:104] = to_bytes_legacy(A, 32, 4, False)
        pd[104:136] = to_bytes_legacy(B, 32, 4, False)
        pd[136:176] = to_bytes_legacy(K, 40, 4, False)
        return from_bytes_legacy(_H(bytes(pd)), 1, False)

    @staticmethod
    def server_password_proof(A: int, M1: int, K: int) -> int:
        pd = bytearray(92)
        pd[0:32] = to_bytes_legacy(A, 32, 4, False)
        pd[32:52] = to_bytes_legacy(M1, 20, 4, False)
        pd[52:92] = to_bytes_legacy(K, 40, 4, False)
        return from_bytes_legacy(_H(bytes(pd)), 1, False)


# --- wire (de)serialisation helpers used by the differential client ----------
def salt_to_wire(salt_int: int) -> bytes:
    """32-byte salt as the client sends it at account creation.

    The server decodes the stored salt with ``BigInt(account_salt, 32, 4,
    false)`` (== ``from_bytes_legacy(.,4,False)``) and then derives the raw
    salt it hashes with as ``getData(buf, 32)`` (block 1, big-endian) — the
    same block-1-big-endian raw salt the client uses internally. For both
    sides to agree on the salt *integer* (and therefore the raw salt, x, and
    M1), the wire bytes must be the inverse of the server's block-4 decode,
    which is ``to_bytes_legacy(.,4,True)`` (verified: round-trips exactly).
    Using block-4 *little*-endian here silently desynchronises the salt and
    makes every proof fail with response 2."""
    return to_bytes_legacy(salt_int, 32, 4, True)


def salt_from_wire(b: bytes) -> int:
    return from_bytes_legacy(b, 4, False)


def verifier_to_wire(v: int) -> bytes:
    """Verifier + A are read by the server with from_bytes_legacy(.,1,false)
    i.e. plain little-endian, so the client serialises them little-endian."""
    return v.to_bytes(32, "little")


def pubkey_A_to_wire(a_pub: int) -> bytes:
    return a_pub.to_bytes(32, "little")


def pubkey_B_from_wire(b: bytes) -> int:
    """Server sends B via to_bytes_legacy(.,4,false) == little-endian bytes."""
    return int.from_bytes(b, "little")


def proof_to_wire(m: int) -> bytes:
    return to_bytes_legacy(m, 20, 4, False)


# --- golden self-check -------------------------------------------------------
def _selfcheck() -> int:
    user, pw = "GOLDEN", "correcthorse"
    salt = int("00112233445566778899aabbccddeeff"
               "0123456789abcdef0123456789abcdef", 16)
    a = int("a1a2a3a4a5a6a7a8a9aaabacadaeafb0"
            "b1b2b3b4b5b6b7b8b9babbbcbdbebfc0", 16)
    b = int("0102030405060708090a0b0c0d0e0f10"
            "1112131415161718191a1b1c1d1e1f20", 16)

    c = BnetSrp3(user, pw)
    c.set_salt(salt)
    c.set_client_private_key(a)

    v = c.verifier()
    A = c.client_session_public_key()
    B = (v + pow(G, b, N)) % N
    K = c.hashed_client_secret(B)
    M1 = c.client_password_proof(A, B, K)
    M2 = c.server_password_proof(A, M1, K)

    expect = {
        "v":  "29af1b9c11c85cc6dce4b6e4b649c0b4af9edccb16b2ac6c8a06d94ad38dd806",
        "A":  "ff153b11871a7d3aa77edd3f1575c2b4fbad0d550e6df0222ccb0b7eb93952b",
        "B":  "4643f16f0cdd83558b6881281a67397db8f635c80160d7b4dca1d99592b6d3fc",
        "K":  "82472ed3e115f06ba57a8ac31683a8464aa1a9c3902b3c003eff58608d9c9876"
              "524cf02f0b0e0d4b",
        "M1": "9454d79ec5c50593b27ad845847a0b9b4551b2fb",
        "M2": "ea80de605bc99538ae64f0ec72299acd6e71b53",
    }
    got = {"v": v, "A": A, "B": B, "K": K, "M1": M1, "M2": M2}
    ok = True
    for name, hexv in expect.items():
        if got[name] != int(hexv, 16):
            ok = False
            print(f"FAIL {name}: got {got[name]:x}\n         exp {hexv}")
    if ok:
        print("PASS: Python SRP-3 reproduces all C++ golden vectors")
        return 0
    return 1


if __name__ == "__main__":
    import sys
    sys.exit(_selfcheck())
